src/lang/defs/finder.rs

```diff
@@ -312,6 +327,71 @@ fn try_variable_declaration(
     None
 }
 
+/// Resolves concrete type identifiers in type-level expressions and type annotations.
+/// In particular, handles **type aliases** which require special care.
+fn try_concrete_type(
+    db: &AnalysisDatabase,
+    identifier: &ast::TerminalIdentifier,
+    lookup_items: &[LookupItemId],
+) -> Option<ResolvedItem> {
+    let ptr = identifier.stable_ptr(db);
+    let module_file_id = db.find_module_file_containing_node(&identifier.as_syntax_node())?;
+
+    for &lookup_item_id in lookup_items {
+        let resolved_generic = db.lookup_resolved_generic_item_by_ptr(lookup_item_id, ptr);
+
+        // If the type obviously resolves to an alias, return immedietely.
+        // This happens only in the `use` path.
+        if let Some(item @ ResolvedGenericItem::GenericTypeAlias(_)) = resolved_generic {
+            return Some(ResolvedItem::Generic(item));
+        }
+
+        // Compiler resolves the type aliases recursively, until it reaches a concrete type.
+        // If we calle [`SemanticGroup::resolve_concrete_item_by_ptr`],
+        // on the identifier, we will receive the aliased type instead of the alias.
+        //
+        // To avoid this, we try resolving the identifier as a path, which is handled differently by the resolver.
+        // This allows us to reach the definition of the alias.
+
+        let type_path_segments =
+            identifier.as_syntax_node().ancestor_of_type::<ast::ExprPath>(db)?.to_segments(db);
+
+        if type_path_segments
+            .last()
+            .is_none_or(|segment| segment.identifier(db) != identifier.text(db))
+        {
+            return None;
+        }
+
+        let mut resolver =
```

Comment:
Shouldn't this be somehow derived from `resolver_data` (from calling function of `try_concrete_type` and passed down to `try_concrete_type`)?